### ISSUE TITLE
IPv4Address (and friends) should override __hash__

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4783]
|<img alt="rfw's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| rfw reported|
|-|-|
|Trac ID|trac#4783|
|Type|enhancement|
|Created|2010-12-29 08:59:50Z|
|Branch|https://github.com/twisted/twisted/tree/address-hash-4783|

In my opinion, `IPv4Address` should implement a `__hash__` function for use in dicts, sets and such.

Attachments:

* [addresshash-4783.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/52a/52a3ba38f88f880b82fe97e4d5a7a9feca3cfa24/ee855210880ef4cad98392f6c56b433221698333.patch) (3233 bytes) - added by michaelnt on 2011-01-29 19:17:37Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4783 4783
type__enhancement enhancement
reporter__rfw rfw
priority__low low
milestone__ 
branch__branches_address_hash_4783 branches/address-hash-4783
branch_author__michaelnt michaelnt
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1293613190000000 1293613190000000
changetime__1297109885000000 1297109885000000
version__None None
owner__ 
cc__awclin cc__itamar
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to rfw|
|-|-|

I suppose that would be fine.  Would you like to [ReviewProcess contribute a patch]?   [#4613](https://github.com/twisted/twisted/issues/4613) may be of interest as well.

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|michaelnt removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

This almost looks good, but I just realized the `os.path.samefile` usage in `UNIXAddress` makes it more complicated.  A `UNIXAddress` using a symlink to a real UNIX socket will compare equal to a `UNIXAddress` using the UNIX socket directly.

Can a `UNIXAddress` ever actually refer to a symlink though?  You cannot bind an AF_UNIX socket to a symlink that exists.

Another way to get `samefile` to return `True` is with hard links.  You cannot bind to an AF_UNIX socket to a socket that exists either, though.

`UNIXAddress` was introduced in r10314 as part of the resolution for [#7200](https://github.com/twisted/twisted/issues/7200).  The `samefile` check was present in the first version of the class, but with no tests or explanations that I can find.  I think itamar wrote it.  _Maybe_ he remembers why it's there...

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

I do not remember. Possibly an abspath on creation would be sufficient.

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|michaelnt commented|
|-|-|

From the python documentation

```
 The only required property is that objects which compare equal have the same hash value
```

So that isn't the case here.

Is the __eq__ method doing the right thing?

If so I guess the hash should try and dereference links and return a hash of the file not the link.

**Comment by twisted-trac** [comment]
|<img alt="awclin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|awclin commented|
|-|-|

FYI __eq__ is being replaced as part of [http://twistedmatrix.com/trac/ticket/4817 [#4817](https://github.com/twisted/twisted/issues/4817)]

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|michaelnt commented|
|-|-|

Updated patch so that `UNIXAddress` objects that compare as equal return the same hash.

As [#4817](https://github.com/twisted/twisted/issues/4817) won't add a hash method I think this change should still go in.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [30639]) Branching to 'address-hash-4783'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [30640]) Apply addresshash-4783.patch

refs [#4783](https://github.com/twisted/twisted/issues/4783)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

The `__hash__` methods look good (although I still don't know if it's _really_ possible to have a UNIX socket via a symlink).  A few minor coding standard points:

  1. Methods should be separated by two blank lines
  1. Lines shouldn't be 80 cols or wider
  1. No `os.symlink` on Windows, so the symlink tests need to skip there.

Nothing too major there, I'll fix and commit.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [30641]) Fix whitespace and line length issues.  Also skip symlink tests on Windows.

refs [#4783](https://github.com/twisted/twisted/issues/4783)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [30642]) Merge address-hash-4783

Author: michaelnt
Reviewer: exarkun
Fixes: [#4783](https://github.com/twisted/twisted/issues/4783)


Make `IPv4Address` and `UNIXAddress` hashable in a way which is compatible
with their already-defined equality semantics.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
